README.md
# kenzy_image &middot; [![GitHub license](https://img.shields.io/github/license/lnxusr1/image_analyzer)](https://github.com/lnxusr1/kenzy_image/blob/master/LICENSE) ![Python Versions](https://img.shields.io/pypi/pyversions/yt2mp3.svg) ![Read the Docs](https://img.shields.io/readthedocs/kenzy_image) ![GitHub release (latest by date)](https://img.shields.io/github/v/release/lnxusr1/kenzy_image)

This module is dedicated to simplifying the interactions required for face detection, face recognition, object detection, and motion detection.  Visit our main site: [https://kenzy.ai/](https://kenzy.ai/)

More info available in the [documentation](https://image-docs.kenzy.ai/)

## Installation

The easiest way to install kenzy_image is with the following:

```
pip install kenzy-image
```

Just make sure you're running Python 3.6 or newer.

## Embedding into your program

Visit the [detector](https://image-docs.kenzy.ai/en/latest/detector/) page in the [documentation](https://image-docs.kenzy.ai/) for more information.

```
import cv2  # For webcam input
from kenzy_image import detector

# Create our Kenzy Detector Object
myImageDetector = detector(detectFaces=True, 
                           detectObjects=True, 
                           objModelType="ssd",  # or yolo
                           detectMotion=True, 
                           imageMarkupFaces=True, 
                           scaleFactor=0.5)

# Add Named Faces to Recognize
myImageDetector.addFace("IMG_4291_portrait_jon.jpg", "Jon Doe")
myImageDetector.addFace("IMG_7033_portrait_jane.jpg", "Jane Doe")

# Open the camera stream
cam = cv2.VideoCapture(0)

while True:

    # Read frame from camera
    ret, frame = cam.read()

    # Analyze with kenzy_image
    myImageDetector.analyze(frame)

    print("=======================")
    print("FACES     =", len(myImageDetector.faces))
    print("OBJECTS   =", [x.get("name") for x in myImageDetector.objects])
    print("MOVEMENTS =", True if len(myImageDetector.movements) > 0 else False)

    # Show the frame with markup
    cv2.imshow('KENZY_IMAGE', myImageDetector.image)

    # Loop until [Esc] or "q"
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cam.release()
cv2.destroyAllWindows()
```

## Running as module

Options are as follows for starting kenzy_image:

```
python -m kenzy_image [OPTIONS]
```

For example, to switch the model to MobileNet v3's SSD model use:
```
python -m kenzy_image --object-detect-type ssd  
```

Use ```--help``` for more information on options.

## References and Other Useful Links

Many thanks to those that build the models and core libraries that KENZY_IMAGE incorporates.  Please find links to those below:

#### Face Detection & Recognition

- [Face Recognition Concepts and Examples](https://git.ece.iastate.edu/se_329_cylicon_valley/face_recognition)
- [Face Recognition Library](https://github.com/ageitgey/face_recognition)

#### Object Detection

- [COCO Inference Model &amp; Config](https://github.com/zafarRehan/object_detection_COCO)
- [COCO Labels](https://tech.amikelive.com/node-718/what-object-categories-labels-are-in-coco-dataset/) <-- Better label set for frozen inference model
- [YOLOv7 Model](https://github.com/wongkinyiu/yolov7) <-- Source for the included [Yolov7-tiny.pt](https://github.com/WongKinYiu/yolov7/releases)
- [YOLOv7 Python Library](https://github.com/akashAD98/yolov7-pip-1)

### Motion Detection

- [Motion Detection Example](https://towardsdatascience.com/image-analysis-for-beginners-creating-a-motion-detector-with-opencv-4ca6faba4b42)

-----

## Help &amp; Support
Help and additional details is available at [https://kenzy.ai](https://kenzy.ai)

Read the docs: [https://image-docs.kenzy.ai/](https://image-docs.kenzy.ai/)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.0.0]

### Added

- ```imageMarkupFaces``` set to ```True``` or ```False``` to indicate showing bounding boxes on detected faces (overrides ```imageMarkup``` value)
- ```imageMarkupObjects``` set to ```True``` or ```False``` to indicate showing bounding boxes on detected objects (overrides ```imageMarkup``` value)
- ```imageMarkupMotion``` set to ```True``` or ```False``` to indicate showing bounding boxes on detected motion (overrides ```imageMarkup``` value)

### Modified

- Corrected default value for ```objModelType``` set to ```ssd``` for direct module execution

## Builds [0.0.6] thru [0.0.9]

### Added

- ```detector.analyze()``` now supports overrides to core setting by supplying the boolean values for detectFaces, detectObjects, and/or detectMotion directly to this function.
- ```objDetectList``` now enables you to specify a list of objects to look for (must be one of the values in the object labels list).
- Enabled ```--camera-device``` for cli calls to specify what device or RTSP stream to use as the source.
- ```objModelType``` now enables `yolo` or `ssd` options for model usage.  (Yolo leverages yolov7-tiny)

### Modified

- Default object detection model changed from MobileNet V3 SSD to Yolov7
- Fixed spelling mistakes in documentation for option variables
- Optimized hierarchy for image processing to improve overall performance and lag
- Corrected error in `--no-faces` option that was preventing it from taking affect
- Corrected error in bounding boxes when scale factor is larger than 1.0 (which is not recommended)
- Corrected bad local import in main init()
- Disabled optimization for processing individual faces rather than entire image due to multi-channel image issues

## [0.0.5]

### Added

- ```orientation``` option with values of 0, 90, 180, or 270 to rotate the input image by the desired degree.

### Modified

- Changed dependency from opencv-python to opencv-contrib-python (Raspberry Pi requires this for Face/Object detection)
- Cleaned up mistake in Detector documentation with invalid object name in example

## [0.0.4]

### Modified

- Fixed but in rt_secs/rtSets for runtime stats.


## [0.0.3]

### Modified

- Multiple updates to documentation
- Corrected a bug when setting font and outline colors using tuples for running as module

## [0.0.2]

This is the initial version and has not be heavily validated or tested.  Stable versions should be available in v1.0 or later.

### Added

- CLI Parameters to enable calling module directly (python -m kenzy_image)
- Default to enable all detections (Face, Object, and Motion)
- Arguments added to enable faces to be recognized e.g. ```--faces /path/image1.jpg LNXUSR1```  (Can add multiple ```--faces``` arguments.)
- Added ```--config``` to enable specifying a JSON formatted file with all options configured.  (See docs for more info.)
- Use ```--help``` for full option list

### Modified

- Renamed module to kenzy_image

src/kenzy_image/__main__.py
import argparse
import logging
import os
import json
import cv2
import time
from . import __app_name__, __version__
from .core import detector


def doParseArg(argName, argValue, cfg):
    if argValue is not None:
        cfg[argName] = argValue

    return


def getTupleValue(inValue):
    if inValue is None or str(inValue).strip() == "":
        return None
    
    if str(inValue).startswith("(") and str(inValue).endswith(")"):
        try:
            return eval(inValue)
        except Exception:
            logger = logging.getLogger("STARTUP")
            logger.warning("Unable to convert value to tuple - " + str(inValue))
            return None
    
    logger = logging.getLogger("STARTUP")
    logger.warning("Value does not appear to be a tuple - " + str(inValue))
    return None


def doParseArgs(cfg, ARGS):
    if not isinstance(cfg, dict):
        cfg = {}

    doParseArg("orientation", ARGS.orientation, cfg)

    if ARGS.no_faces:
        cfg["detectFaces"] = False
    
    if ARGS.no_objects:
        cfg["detectObjects"] = False
    
    if ARGS.no_motion:
        cfg["detectMotion"] = False

    if ARGS.no_markup:
        cfg["imageMarkup"] = False

    doParseArg("scaleFactor", ARGS.scale_factor, cfg)

    # Faces
    doParseArg("defaultFaceName", ARGS.face_detect_default_name, cfg)
    doParseArg("faceModel", ARGS.face_detect_model, cfg)
    doParseArg("faceFontColor", getTupleValue(ARGS.face_detect_font_color), cfg)
    doParseArg("faceOutlineColor", getTupleValue(ARGS.face_detect_outline_color), cfg)

    if ARGS.no_face_names:
        cfg["showFaceNames"] = False

    # Objects
    doParseArg("objModelType", ARGS.object_detect_type, cfg)
    doParseArg("objDetectCfg", ARGS.object_detect_config, cfg)
    doParseArg("objDetectList", ARGS.object_list, cfg)
    doParseArg("objDetectModel", ARGS.object_detect_model, cfg)
    doParseArg("objDetectLabels", ARGS.object_detect_labels, cfg)
    doParseArg("objFontColor", getTupleValue(ARGS.object_detect_font_color), cfg)
    doParseArg("objOutlineColor", getTupleValue(ARGS.object_detect_outline_color), cfg)

    if ARGS.no_object_names:
        cfg["showObjectNames"] = False

    # Motion
    doParseArg("motionThreshold", ARGS.motion_detect_threshold, cfg)
    doParseArg("motionMinArea", ARGS.motion_detect_min_area, cfg)
    doParseArg("motionOutlineColor", getTupleValue(ARGS.motion_detect_outline_color), cfg)

    return cfg


parser = argparse.ArgumentParser(
    description=__app_name__ + " v" + __version__,
    formatter_class=argparse.RawTextHelpFormatter,
    epilog='''To start the services try:\npython3 -m kenzy_image\n\nMore information available at:\nhttp://kenzy.ai''')

parser.add_argument('-c', '--config', default=None, help="Configuration file")
parser.add_argument('-v', '--version', action="store_true", help="Print Version")

startup_group = parser.add_argument_group('Startup Options')

startup_group.add_argument('--camera-device', default=None, help="Camera object to use")
startup_group.add_argument('--no-markup', action="store_true", help="Hide outlines and names")
startup_group.add_argument('--scale-factor', default=None, help="Image scale factor (decimal).  Values < 1 improve performance.")
startup_group.add_argument('--orientation', default=None, help="Image orientation (0, 90, 180, or 270)")

face_group = parser.add_argument_group('Face Detection')

face_group.add_argument('--no-faces', action="store_true", help="Disable face detection")
face_group.add_argument('--face-detect-default-name', default=None, help="Set the Unknown face name")
face_group.add_argument('--face-detect-model', default=None, help="Model to leverage (hog or cnn)")
face_group.add_argument('--face-detect-font-color', default=None, help="Face names font color as tuple e.g. (0, 0, 255)")
face_group.add_argument('--face-detect-outline-color', default=None, help="Faces outline color as tuple e.g. (0, 0, 255)")
face_group.add_argument('--no-face-names', action="store_true", help="Hides the face names even if identified.")
face_group.add_argument('--faces', action="append", nargs=2, metavar=("path", "name"), help="Face image and name e.g. --face image.jpg LNXUSR1")

object_group = parser.add_argument_group('Object Detection')

object_group.add_argument('--no-objects', action="store_true", help="Disable object detection")
object_group.add_argument('--object-detect-type', default="ssd", help="Object detection type (yolo or ssd)")
object_group.add_argument('--object-detect-config', default=None, help="Object detection configuration")
object_group.add_argument('--object-detect-model', default=None, help="Object detection inference model file")
object_group.add_argument('--object-detect-labels', default=None, help="Object detection inference model label files")
object_group.add_argument('--object-detect-font-color', default=None, help="Object names font color as tuple e.g. (0, 0, 255)")
object_group.add_argument('--object-detect-outline-color', default=None, help="Object detection outline color as tuple e.g. (0, 0, 255)")
object_group.add_argument('--object-list', default=None, help="Limit list of objects to detect detection (optional)")
object_group.add_argument('--no-object-names', action="store_true", help="Hides the object names even if identified.")

motion_group = parser.add_argument_group('Motion Detection')

motion_group.add_argument('--no-motion', action="store_true", help="Disable motion detection")
motion_group.add_argument('--motion-detect-threshold', default=None, help="Motion detection difference threshold")
motion_group.add_argument('--motion-detect-min-area', default=None, help="Motion detection minimum pixel area")
motion_group.add_argument('--motion-detect-outline-color', default=None, help="Motion area outline color as tuple e.g. (0, 0, 255)")

logging_group = parser.add_argument_group('Logging Options')

logging_group.add_argument('--log-level', default="info", help="Options are full, debug, info, warning, error, and critical")
logging_group.add_argument('--log-file', default=None, help="Redirects all logging messages to the specified file")

ARGS = parser.parse_args()

if ARGS.version:
    print(__app_name__, "v" + __version__)
    quit()

logLevel = logging.INFO
if ARGS.log_level is not None and ARGS.log_level.strip().lower() in ["debug", "info", "warning", "error", "critical"]:
    logLevel = eval("logging." + ARGS.log_level.strip().upper())
elif ARGS.log_level is not None and ARGS.log_level.strip().lower() == "full":
    logLevel = logging.DEBUG

for handler in logging.root.handlers[:]:
    logging.root.removeHandler(handler)

logging.basicConfig(
    datefmt='%Y-%m-%d %H:%M:%S %z',
    filename=ARGS.log_file,
    format='%(asctime)s %(name)-12s - %(levelname)-9s - %(message)s',
    level=logLevel)

logger = logging.getLogger("STARTUP")

cfg = {}
if ARGS.config is not None and os.path.isfile(ARGS.config):
    with open(ARGS.config, "r", encoding="UTF-8") as fp:
        cfg = json.load(fp)

cfg = doParseArgs(cfg, ARGS)
obj = detector(**cfg)

if ARGS.faces is not None and isinstance(ARGS.faces, list):
    for item in ARGS.faces:
        if not os.path.isfile(item[0]):
            raise FileNotFoundError("Could not find file: " + str(item[0]))

        obj.addFace(item[0], item[1])

devId = 0
if ARGS.camera_device is not None:
    try:
        devId = int(str(ARGS.camera_device).strip())
    except ValueError:
        devId = str(ARGS.camera_device).strip()

logger = logging.getLogger("RESULTS")
cam = cv2.VideoCapture(devId)
while True:

    start = time.time()
    
    ret, frame = cam.read()
    obj.analyze(frame)

    end = time.time()

    logger.info("=======================")
    logger.info("FACES      = " + str(len(obj.faces)))
    logger.info("OBJECTS    = " + str([x.get("name") for x in obj.objects]))
    logger.info("MOVEMENTS  = " + str(True if len(obj.movements) > 0 else False))
    logger.info("TOTAL TIME = " + str(end - start) + " seconds")

    cv2.imshow('KENZY_IMAGE', obj.image)

    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cam.release()
cv2.destroyAllWindows()